zigzag-circuit.md

```diff
@@ -0,0 +1,122 @@
+# ZigZag: Offline PoRep Circuit Spec
+
+ZigZag is the Proof of Replication used in Filecoin. The prover encodes the original data into a replica and commits to it. An offline PoRep proves that the commitment to the replica is a valid commitment of the encoded original data.
+
+#### ZigZag Overview
+
+ZigZag is the Proof of Replication used in Filecoin. The prover encodes the original data into a replica and commits to it. An offline PoRep proves that the commitment to the replica is a valid commitment of the encoded original data.
+
+ZigZag PoRep is based on layering DRG graphs `l` times. The data represented in each DRG layer is the data encoded in the previous layer. The final layer is the replica (which in Filecoin terms is the sealed sector).
+
+- `replicaId` is a unique replica identifier (see the Filecoin Proofs spec for details)
+- `CommD` is the Merkle Tree root hash of the input data to the first layer
+- `CommR_{i}` is the Merkle Tree hash of the output of the DRG encoding at each layer `i` 
+- `CommRStar` is the hash of the concatenation of the `replicaId` and all the `commR`s.
+
+In the PoRep offline proof as presented in Fisch2019, the proof is large for blockchain usage. We use SNARKs to generate a proof of knowledge of an offline proof. In other words, we use encode the ZigZag proof verification algorithm in an arithmetic circuit and use SNARKs to prove that it was evaluated correctly.
+
+This circuit proves that given a Merkle root `CommD`, `CommR_l`, and `commRStar`, that the prover knew the correct replicated data at each layer.
+
+## Offline PoRep circuit
+
+**Public Parameters**: *Parameters that are embeded in the circuits or used to generate the circuit*
+
+- `LAYERS : UInt`: Number of layers
+- `LAYER_CHALLENGES[0..LAYERS]`: Number of challenges per layer
+- `EXPANSION_DEGREE: UInt`: Degree of the bipartite expander graph to extend dependencies between layers
+- `TREE_DEPTH: UInt`: Height of the Merkle tree
+
+**Public Inputs**: *Inputs that the prover uses to generate a SNARK proof and that the verifier uses to verify it*
+
+- `ReplicaId : Fr`: A unique identifier for the replica Id
+- `CommD : Fr`: the Merkle tree root hash of the original data (input to the first layer)
+- `CommRLast : Fr`: the Merkle tree root hash of the output of the last layer 
+- `CommRStar : Fr`: the aggregate of each layer's Merkle tree root hash
+- Inclusion paths: Binary representation of the Merkle tree path that must be proven packed into a single `Fr` element. We have the following inclusion paths:
+  - `InclusionPaths_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`: At each layer `i` we have `LAYER_CHALLENGES[i]` inclusion paths.
+  - `ParentsInclusionPaths_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}`: At each layer `l` we have `LAYER_CHALLENGES[i]` an inclusion path for each parent node of the corresponding `InclusionPaths_{l}_{c}`.
```

Comment:
nit: `l=0` -> `i=0` (same for the other `l` references, we seem to use that letter for the total and `i` for each one)

Reply:
fixing